[Bug]: `draw_all_possible_flows` doesn't include return event types as nodes

### Bug Description

Title says everything. I filled relevant sections for buggy example and output.

I was puzzled for quite a time. And here is what I found:

At <https://github.com/run-llama/llama_index/blob/main/llama-index-utils/llama-index-utils-workflow/llama_index/utils/workflow/draw.py#L61>:

```python
...
        for return_type in step_config.return_types:
            print(f"{return_type}", flush=True)
            if return_type == type(None):
                continue

            if issubclass(return_type, InputRequiredEvent):
                net.add_node(
                    return_type.__name__,
                    label=return_type.__name__,
                    color="#90EE90",
                    shape="ellipse",
                )  # Light green for events

                if not net_has_external_step_node:
                    # add node for conceptual external step
                    net.add_node(
                        f"external_step",
                        label="external_step",
                        color="#BEDAE4",
                        shape="box",
                    )
...
```

`draw_all_possible_flows` doesn't include return types as nodes. And thus, my custom `StopEvent` couldn't be shown.




### Version

0.12.25

### Steps to Reproduce

Buggy example:

```python
import asyncio
from llama_index.core.workflow import StartEvent, StopEvent, Workflow, draw_all_possible_flows, step


class MyStopEvent(StopEvent):
    pass


class TestWorkflow(Workflow):
    @step
    async def start(self, ev: StartEvent) -> MyStopEvent:
        return MyStopEvent()


async def main():
    draw_all_possible_flows(TestWorkflow(), "output.html")


asyncio.run(main())
```


### Relevant Logs/Tracbacks

```shell
<class 'NoneType'>
<class '__main__.MyStopEvent'>
Traceback (most recent call last):
  File "/mnt/data/projects/ai-paper-gen/workflows/test.py", line 20, in <module>
    asyncio.run(main())
    ~~~~~~~~~~~^^^^^^^^
  File "/usr/lib/python3.13/asyncio/runners.py", line 195, in run
    return runner.run(main)
           ~~~~~~~~~~^^^^^^
  File "/usr/lib/python3.13/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/usr/lib/python3.13/asyncio/base_events.py", line 725, in run_until_complete
    return future.result()
           ~~~~~~~~~~~~~^^
  File "/mnt/data/projects/ai-paper-gen/workflows/test.py", line 17, in main
    draw_all_possible_flows(TestWorkflow(), "output.html")
    ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/data/projects/ai-paper-gen/.venv/lib/python3.13/site-packages/llama_index/utils/workflow/draw.py", line 92, in draw_all_possible_flows
    net.add_edge(step_name, return_type.__name__)
    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/data/projects/ai-paper-gen/.venv/lib/python3.13/site-packages/pyvis/network.py", line 372, in add_edge
    assert to in self.get_nodes(), \
           ^^^^^^^^^^^^^^^^^^^^^^
AssertionError: non existent node 'MyStopEvent'
```

--- a/llama-index-utils/llama-index-utils-workflow/llama_index/utils/workflow/draw.py
+++ b/llama-index-utils/llama-index-utils-workflow/llama_index/utils/workflow/draw.py
@@ -33,7 +33,7 @@ def draw_all_possible_flows(
     step_config: Optional[StepConfig] = None
 
     # Only one kind of `StopEvent` is allowed in a `Workflow`.
-    # Assuming that `Workflow` is vaidated before drawing, it's enough to find the first one.
+    # Assuming that `Workflow` is validated before drawing, it's enough to find the first one.
     current_stop_event = None
     for step_name, step_func in steps.items():
         step_config = getattr(step_func, "__step_config", None)
@@ -60,7 +60,7 @@ def draw_all_possible_flows(
         for event_type in step_config.accepted_events:
             if event_type == StopEvent and event_type != current_stop_event:
                 continue
-            
+
             net.add_node(
                 event_type.__name__,
                 label=event_type.__name__,
@@ -78,7 +78,7 @@ def draw_all_possible_flows(
                 color="#90EE90",
                 shape="ellipse",
             )  # Light green for events
-                
+
             if issubclass(return_type, InputRequiredEvent):
                 # add node for conceptual external step
                 net.add_node(

--- a/llama-index-utils/llama-index-utils-workflow/llama_index/utils/workflow/draw.py
+++ b/llama-index-utils/llama-index-utils-workflow/llama_index/utils/workflow/draw.py
@@ -20,7 +20,7 @@ def draw_all_possible_flows(
     filename: str = "workflow_all_flows.html",
     notebook: bool = False,
 ) -> None:
-        """Draws all possible flows of the workflow."""
+    """Draws all possible flows of the workflow."""
     from pyvis.network import Network
 
     net = Network(directed=True, height="750px", width="100%")

--- a/llama-index-utils/llama-index-utils-workflow/llama_index/utils/workflow/draw.py
+++ b/llama-index-utils/llama-index-utils-workflow/llama_index/utils/workflow/draw.py
@@ -1,4 +1,3 @@
-
 from typing import Optional
 
 from llama_index.core.workflow.events import (

--- a/llama-index-utils/llama-index-utils-workflow/llama_index/utils/workflow/draw.py
+++ b/llama-index-utils/llama-index-utils-workflow/llama_index/utils/workflow/draw.py
@@ -1,3 +1,4 @@
+
 from typing import Optional
 
 from llama_index.core.workflow.events import (
@@ -19,19 +20,10 @@ def draw_all_possible_flows(
     filename: str = "workflow_all_flows.html",
     notebook: bool = False,
 ) -> None:
-    """Draws all possible flows of the workflow."""
+        """Draws all possible flows of the workflow."""
     from pyvis.network import Network
 
     net = Network(directed=True, height="750px", width="100%")
-
-    # Add the nodes + edge for stop events
-    net.add_node(
-        StopEvent.__name__,
-        label=StopEvent.__name__,
-        color="#FFA07A",
-        shape="ellipse",
-    )
-    net.add_node("_done", label="_done", color="#ADD8E6", shape="box")
 
     # Add nodes from all steps
     steps = get_steps_from_class(workflow)
@@ -40,7 +32,23 @@ def draw_all_possible_flows(
         steps = get_steps_from_instance(workflow)
 
     step_config: Optional[StepConfig] = None
-    net_has_external_step_node: bool = False
+
+    # Only one kind of `StopEvent` is allowed in a `Workflow`.
+    # Assuming that `Workflow` is vaidated before drawing, it's enough to find the first one.
+    current_stop_event = None
+    for step_name, step_func in steps.items():
+        step_config = getattr(step_func, "__step_config", None)
+        if step_config is None:
+            continue
+
+        for return_type in step_config.return_types:
+            if issubclass(return_type, StopEvent):
+                current_stop_event = return_type
+                break
+
+        if current_stop_event:
+            break
+
     for step_name, step_func in steps.items():
         step_config = getattr(step_func, "__step_config", None)
         if step_config is None:
@@ -51,6 +59,9 @@ def draw_all_possible_flows(
         )  # Light blue for steps
 
         for event_type in step_config.accepted_events:
+            if event_type == StopEvent and event_type != current_stop_event:
+                continue
+            
             net.add_node(
                 event_type.__name__,
                 label=event_type.__name__,
@@ -59,26 +70,24 @@ def draw_all_possible_flows(
             )  # Light green for events
 
         for return_type in step_config.return_types:
-            print(f"{return_type}", flush=True)
             if return_type == type(None):
                 continue
 
+            net.add_node(
+                return_type.__name__,
+                label=return_type.__name__,
+                color="#90EE90",
+                shape="ellipse",
+            )  # Light green for events
+                
             if issubclass(return_type, InputRequiredEvent):
+                # add node for conceptual external step
                 net.add_node(
-                    return_type.__name__,
-                    label=return_type.__name__,
-                    color="#90EE90",
-                    shape="ellipse",
-                )  # Light green for events
-
-                if not net_has_external_step_node:
-                    # add node for conceptual external step
-                    net.add_node(
-                        f"external_step",
-                        label="external_step",
-                        color="#BEDAE4",
-                        shape="box",
-                    )
+                    f"external_step",
+                    label="external_step",
+                    color="#BEDAE4",
+                    shape="box",
+                )
 
     # Add edges from all steps
     for step_name, step_func in steps.items():
@@ -95,7 +104,10 @@ def draw_all_possible_flows(
                 net.add_edge(return_type.__name__, f"external_step")
 
         for event_type in step_config.accepted_events:
-            net.add_edge(event_type.__name__, step_name)
+            if step_name == "_done" and issubclass(event_type, StopEvent):
+                net.add_edge(current_stop_event.__name__, step_name)
+            else:
+                net.add_edge(event_type.__name__, step_name)
 
             if issubclass(event_type, HumanResponseEvent):
                 net.add_edge(

--- a/llama-index-utils/llama-index-utils-workflow/pyproject.toml
+++ b/llama-index-utils/llama-index-utils-workflow/pyproject.toml
@@ -24,7 +24,7 @@ exclude = ["**/BUILD"]
 license = "MIT"
 name = "llama-index-utils-workflow"
 readme = "README.md"
-version = "0.3.0"
+version = "0.3.1"
 
 [tool.poetry.dependencies]
 python = ">=3.9,<4.0"

